src/server/page_server.cpp

```diff
@@ -371,6 +377,79 @@ page_server::async_disconnect_handler::disconnect_loop ()
     }
 }
 
+void page_server::pts_mvcc_tracker::init_oldest_active_mvccid (const std::string &pts_channel_id)
+{
+  std::lock_guard<std::mutex> lockg { m_pts_oldest_active_mvccids_mtx };
+  /*
+   * The entry must not already be present. If the same passive transaction server has been connected
+   * before, the entry must have been removed when the PTS disconnected or when the connection
+   *  to the PTS was aborted.
+   */
+  assert (m_pts_oldest_active_mvccids.find (pts_channel_id) == m_pts_oldest_active_mvccids.end());
+
+  /*
+   * MVCCID_ALL_VISIBLE means that it hasn't yet received. It will prevent the ATS to run vacuum.
+   * This is a guard for the window in which a PTS is connected but has't sent its oldest active mvccid.
+   * In this window, if we vaccum without considering the PTS, we possibly end up cleaning up the data
+   * a read-only transaction on the PTS see.
+   */
+  m_pts_oldest_active_mvccids[pts_channel_id] = MVCCID_ALL_VISIBLE;
+}
+
+void page_server::pts_mvcc_tracker::update_oldest_active_mvccid (const std::string &pts_channel_id, const MVCCID mvccid)
+{
+  std::lock_guard<std::mutex> lockg { m_pts_oldest_active_mvccids_mtx };
+
+  /*
+   * 1. The entry is already created when ths PTS is connected.
+   * 2. It is updated by the PTS only when it move foward.
+   *    Without update, it is MVCCID_ALL_VISIBLE by default, which is lower than any mvccid assigned.
+   */
+  assert (m_pts_oldest_active_mvccids.find (pts_channel_id) != m_pts_oldest_active_mvccids.end());
+  assert (m_pts_oldest_active_mvccids[pts_channel_id] < mvccid);
+
+  m_pts_oldest_active_mvccids[pts_channel_id] = mvccid;
+
+#if !defined(NDEBUG)
+  std::string msg;
+  msg.append ("receive_oldest_active_mvccid: update the oldest active mvccid to " + std::to_string (
+		      mvccid) + " of " + pts_channel_id);
+  msg.append ("\n oldest mvcc ids:");
+  for (const auto &it : m_pts_oldest_active_mvccids)
+    {
+      msg.append (" " + std::to_string (it.second));
+    }
+  er_log_debug (ARG_FILE_LINE, msg.c_str());
+#endif
+}
+void page_server::pts_mvcc_tracker::delete_oldest_active_mvccid (const std::string &pts_channel_id)
+{
+  std::lock_guard<std::mutex> lockg { m_pts_oldest_active_mvccids_mtx };
+  /* The entry is already created when ths PTS is connected. */
+  assert (m_pts_oldest_active_mvccids.find (pts_channel_id) != m_pts_oldest_active_mvccids.end());
+  m_pts_oldest_active_mvccids.erase (pts_channel_id);
+}
+
+MVCCID page_server::pts_mvcc_tracker::get_global_oldest_active_mvccid ()
+{
+  std::lock_guard<std::mutex> lockg { m_pts_oldest_active_mvccids_mtx };
+
+  MVCCID oldest_mvccid = MVCCID_LAST;
+  for (const auto &it : m_pts_oldest_active_mvccids)
+    {
+      if (oldest_mvccid > it.second)
```

Comment:
The condition of if statement is always false. So, This function always returns -1(MVCCID_LAST).

Reply:
The type of MVCCID is UINT64 so that -1 means the max value of the type. 2^64-1.

Reply:
OK, I'll use `std::numeric_limits`to prevent confusion.

Reply:
OK.